src/registry.cr

```diff
@@ -0,0 +1,588 @@
+#
+# The Windows API defines some [predefined root keys](https://docs.microsoft.com/en-us/windows/desktop/sysinfo/predefined-keys) that are always open.
+# They are available as constants and can be used as entry points to the registry.
+#
+# * `HKEY_CLASSES_ROOT`
+# * `HKEY_CURRENT_USER`
+# * `HKEY_LOCAL_MACHINE`
+# * `HKEY_USERS`
+# * `HKEY_CURRENT_CONFIG`
+module Registry
+  # Information about file types and their properties.
+  CLASSES_ROOT = Key.new(LibC::HKEY_CLASSES_ROOT, "HKEY_CLASSES_ROOT")
+
+  # Preferences of the current user.
+  #
+  # The kay maps to the current user's subkey in `HKEY_USERS`.
+  CURRENT_USER = Key.new(LibC::HKEY_CURRENT_USER, "HKEY_CURRENT_USER")
+
+  # Information about the physical state of the computer, including installed hardware and software.
+  LOCAL_MACHINE = Key.new(LibC::HKEY_LOCAL_MACHINE, "HKEY_LOCAL_MACHINE")
+
+  # Preferences of all users.
+  #
+  # This key contains a subkey per user.
+  USERS = Key.new(LibC::HKEY_USERS, "HKEY_USERS")
+
+  # Provides access to performance data.
+  PERFORMANCE_DATA = Key.new(LibC::HKEY_PERFORMANCE_DATA, "HKEY_PERFORMANCE_DATA")
+
+  # Information about the hardware profile of the local computer system.
+  CURRENT_CONFIG = Key.new(LibC::HKEY_CURRENT_CONFIG, "HKEY_CURRENT_CONFIG")
+
+  # Registry error.
+  class Error < Exception
+  end
+
+  # Registry key security and access rights.
+  # See https://msdn.microsoft.com/en-us/library/windows/desktop/ms724878.aspx
+  # for details.
+  alias SAM = LibC::REGSAM
+
+  # Registry value types.
+  alias ValueType = LibC::ValueType
+
+  # Union of Crystal types representing a registry value.
+  alias Value = Bytes | Int32 | String | Int64 | Array(String)
+
+  # This type represents a handle to an open Windows registry key.
+  #
+  # Keys can be obtained by calling `#open` on an already opened key.
+  # The predefined root keys are defined as constants in `Registry`.
+  struct Key
+    @handle : LibC::HKEY
+    @name : String
+
+    # Creates a new instance from an open Windows *handle* (`HKEY`) and *name*.
+    #
+    # NOTE: This method is only useful if a Windows handle is retrieved from
+    # calling external code. Usually, new keys are created by opening subkeys
+    # of the predefined root keys available as constants in `Registry`.
+    def initialize(@handle : LibC::HKEY, @name : String)
+    end
+
+    # Returns the full path of this key.
+    getter name : String
+
+    # Returns the Windows handle (`HKEY`) representing this key.
+    def to_unsafe : LibC::HKEY
+      @handle
+    end
+
+    def get_raw?(name : String, buffer : Slice(UInt8)) : {ValueType, UInt32}?
+      name.check_no_null_byte
+
+      get_raw?(name.to_utf16, buffer)
+    end
+
+    private def get_raw?(name : Slice(UInt16), buffer : Bytes) : {ValueType, UInt32}?
+      length = buffer.size.to_u32
+      status = LibC.RegQueryValueExW(self, name, nil, out valtype, buffer, pointerof(length))
+      case status
+      when WinError::ERROR_SUCCESS
```

Comment:
I just now see this, but this should also probably be named `Windows::Error`. Basically anything that's windows specific must be in the `Windows` module.

Reply:
`ERROR_SUCCESS` should be a `LibC` constant. And `WinError` should eventually be entirely replaced by specific cross-platform error types, same as `Errno` (similar to what was proposed in #5003).

Reply:
Right, that makes sense.